### PR TITLE
[SPARK-37937][SQL] Use error classes in the parsing errors of lateral join

### PR DESCRIPTION
**Pull request by imback82**
### What changes were proposed in this pull request?
In the PR, I propose to use the following error classes for the parsing errors of lateral joins:
- `INVALID_SQL_SYNTAX `
- `UNSUPPORTED_FEATURE `

These new error classes are added to `error-classes.json`.

### Why are the changes needed?

Porting the parsing errors for lateral join to the new error framework should improve user experience with Spark SQL.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added new test suite

### CODE REVIEW

**Comment by imback82** [comment]
cc @MaxGekk @cloud-fan @karenfeng

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -89,13 +89,25 @@
     "message" : [ "invalid input syntax for type numeric: %s. To return NULL instead, use 'try_cast'. If necessary set %s to false to bypass this error." ],
     "sqlState" : "42000"
   },
+  "INVALID_LATERAL_JOIN_RELATION" : {
+    "message" : [ "LATERAL can only be used with subquery." ],
+    "sqlState" : "42000"
```
Comment (imback82) [inline core/src/main/resources/error/error-classes.json:94]: @MaxGekk What's the best way to deduce the sql state for each error?

**Review by MaxGekk — COMMENTED** [review]
@imback82 Could you fix the JIRA id in PR's title. Seems SPARK-37858 is not correct.

**Comment by imback82** [comment]
> @imback82 Could you fix the JIRA id in PR's title. Seems SPARK-37858 is not correct.

Updated, thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryParsingErrorsSuite.scala [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+package org.apache.spark.sql.errors
+
+import org.apache.spark.sql.QueryTest
+import org.apache.spark.sql.catalyst.parser.ParseException
+import org.apache.spark.sql.test.SharedSparkSession
+
+class QueryParsingErrorsSuite extends QueryTest with SharedSparkSession {
+  def validateParsingError(
+    sqlText: String,
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryParsingErrorsSuite.scala:26]: nit: 4 spaces indentation

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryParsingErrorsSuite.scala [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      ("LEFT ANTI", "LeftAnti")).foreach { pair =>
+      validateParsingError(
+        sqlText = s"SELECT * FROM t1 ${pair._1} JOIN LATERAL (SELECT c1 + c2 AS c3) ON c2 = c3",
+        errorClass = "UNSUPPORTED_FEATURE",
+        sqlState = "0A000",
+        message = s"The feature is not supported: LATERAL join type '${pair._2}'.")
+    }
+  }
+
+  test("SPARK-35789: INVALID_LATERAL_JOIN_RELATION - LATERAL can only be used with subquery") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryParsingErrorsSuite.scala:67]: I can't really tell if it's an unsupported feature or invalid syntax. Probably the latter as `SELECT * FROM t1, LATERAL (LATERAL t2)` looks really like a bad SQL.

How about we add a new error class `INVALID_SQL_SYNTAX` for this?

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @imback82 and @cloud-fan for review.

**Pull request closed without merging**